ocdata/core/bulk.py

```diff
@@ -0,0 +1,287 @@
+        self.bulk_id_from_db = bulk_id_from_db
+
+    def get_slabs(self):
+        """
+        Returns a list of possible slabs for this bulk instance.
+        This can be later used to iterate through all slabs,
+        or select one at random, to make a Surface object.
+        """
+        if self.precomputed_slabs_path is not None and self.bulk_id_from_db is not None:
+            slabs = self.get_precomputed_slabs()
+        else:
+            slabs = self.compute_slabs()
+
+        return slabs
+
+    def get_precomputed_slabs(self):
+        """
+        Loads relevant pickle of precomputed slabs, and returns
+        a list of tuples containing: (atoms, miller, shift, top).
+        """
+        assert self.bulk_id_from_db is not None
+
+        bulk_slabs_path = os.path.join(
+            self.precomputed_slabs_path, f"{self.bulk_id_from_db}.pkl"
+        )
+        assert os.path.exists(bulk_slabs_path)
+
+        with open(bulk_slabs_path, "rb") as f:
+            slabs_info = pickle.load(f)
+
+        return slabs_info
+
+    def compute_slabs(self, max_miller=MAX_MILLER):
+        """
+        Enumerates all the symmetrically distinct slabs of a bulk structure.
+        It will not enumerate slabs with Miller indices above the
+        `max_miller` argument. Note that we also look at the bottoms of slabs
+        if they are distinct from the top. If they are distinct, we flip the
+        surface so the bottom is pointing upwards.
+
+        Args:
+            max_miller  An integer indicating the maximum Miller index of the slabs
+                        you are willing to enumerate. Increasing this argument will
+                        increase the number of slabs, but the slabs will
+                        generally become larger.
+        Returns:
+            all_slabs_info  A list of 4-tuples containing:  `pymatgen.Structure`
+                            objects for slabs we have enumerated, the Miller
+                            indices, floats for the shifts, and Booleans for "top".
+        """
+        bulk_struct = standardize_bulk(self.atoms)
+
+        all_slabs_info = []
+        for millers in get_symmetrically_distinct_miller_indices(
+            bulk_struct, MAX_MILLER
+        ):
+            slab_gen = SlabGenerator(
+                initial_structure=bulk_struct,
+                miller_index=millers,
+                min_slab_size=7.0,
+                min_vacuum_size=20.0,
+                lll_reduce=False,
+                center_slab=True,
+                primitive=True,
+                max_normal_search=1,
+            )
+            slabs = slab_gen.get_slabs(
+                tol=0.3, bonds=None, max_broken_bonds=0, symmetrize=False
+            )
+
+            # Additional filtering for the 2D materials' slabs
+            if self.mpid is not None and self.mpid in COVALENT_MATERIALS_MPIDS:
+                slabs = [slab for slab in slabs if is_2D_slab_reasonable(slab) is True]
```

Comment:
This is limited to MP. 

@brookwander 
Do we want to think about doing the same for other bulk databases? If `COVALENT_MATERIALS_MPIDS` are materials that are non-metallic favoring covalent bonds can we instead specify a list of COVALENT_ELEMENTS and then perform this check if elements are all COVALENT? i.e.


```
bulk_elems = set(self.atoms.symbols)

# if all bulk elements are covalent elements
if (bulk_elems - COVALENT_ELEMENTS) == 0:
    slabs = [slab for slab in slabs if is_2D_slab_reasonable(slab) is True]
```

Reply:
@mshuaibii oof I started a thread about this as well. I think this is only used to identify 2d materials. I think it makes the most sense to store this in the bulk db so its obvious it needs to be updated? I think we can have a discussion about the 2d material check to because maybe we dont need this as all.